openssl/src/crypto/sign.rs

```diff
@@ -0,0 +1,274 @@
+
+pub struct Signer<'a>(*mut ffi::EVP_MD_CTX, PhantomData<&'a PKey>);
+
+impl<'a> Drop for Signer<'a> {
+    fn drop(&mut self) {
+        unsafe {
+            EVP_MD_CTX_free(self.0);
+        }
+    }
+}
+
+impl<'a> Signer<'a> {
+    pub fn new(type_: Type, pkey: &'a PKey) -> Result<Signer<'a>, ErrorStack> {
+        unsafe {
+            ffi::init();
+
+            let ctx = try_ssl_null!(EVP_MD_CTX_new());
+            let r = ffi::EVP_DigestSignInit(ctx,
+                                            ptr::null_mut(),
+                                            type_.evp_md(),
+                                            ptr::null_mut(),
+                                            pkey.as_ptr());
+            if r != 1 {
+                EVP_MD_CTX_free(ctx);
+                return Err(ErrorStack::get());
+            }
+
+            Ok(Signer(ctx, PhantomData))
+        }
+    }
+
+    pub fn update(&mut self, buf: &[u8]) -> Result<(), ErrorStack> {
+        unsafe {
+            try_ssl_if!(ffi::EVP_DigestUpdate(self.0, buf.as_ptr() as *const _, buf.len()) != 1);
+            Ok(())
+        }
+    }
+
+    pub fn finish(&self) -> Result<Vec<u8>, ErrorStack> {
+        unsafe {
+            let mut len = 0;
+            try_ssl_if!(ffi::EVP_DigestSignFinal(self.0, ptr::null_mut(), &mut len) != 1);
+            let mut buf = vec![0; len];
+            try_ssl_if!(ffi::EVP_DigestSignFinal(self.0, buf.as_mut_ptr() as *mut _, &mut len)
+                    != 1);
+            Ok(buf)
+        }
+    }
+}
+
+impl<'a> Write for Signer<'a> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        try!(self.update(buf));
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+pub struct Verifier<'a>(*mut ffi::EVP_MD_CTX, PhantomData<&'a PKey>);
+
+impl<'a> Drop for Verifier<'a> {
+    fn drop(&mut self) {
+        unsafe {
+            EVP_MD_CTX_free(self.0);
+        }
+    }
+}
+
+impl<'a> Verifier<'a> {
+    pub fn new(type_: Type, pkey: &'a PKey) -> Result<Verifier<'a>, ErrorStack> {
+        unsafe {
+            ffi::init();
+
+            let ctx = try_ssl_null!(EVP_MD_CTX_new());
+            let r = ffi::EVP_DigestVerifyInit(ctx,
+                                              ptr::null_mut(),
+                                              type_.evp_md(),
+                                              ptr::null_mut(),
+                                              pkey.as_ptr());
+            if r != 1 {
+                EVP_MD_CTX_free(ctx);
+                return Err(ErrorStack::get());
+            }
+
+            Ok(Verifier(ctx, PhantomData))
+        }
+    }
+
+    pub fn update(&mut self, buf: &[u8]) -> Result<(), ErrorStack> {
+        unsafe {
+            try_ssl_if!(ffi::EVP_DigestUpdate(self.0, buf.as_ptr() as *const _, buf.len()) != 1);
+            Ok(())
+        }
+    }
+
+    pub fn finish(&self, signature: &[u8]) -> Result<(), ErrorStack> {
```

Comment:
This API does not make sense.  When you are verifying, you need to be able to get the actual verification result.  That is, the return type should be `... -> Result<bool, ErrorStack>` similar to #420 (https://github.com/sfackler/rust-openssl/pull/420/files#diff-35c3c6c147bd45b93218a395c7901036R190).

That is, it needs to be possible to differentiate a true error calculating the signature from the signature failing (probably because either the signature was bad or the contents are corrupt/tampered with).